client/tests/util/test.js

```diff
@@ -50,6 +50,28 @@ test.beforeEach(t => {
     builder = builder
       .forBrowser("chrome")
       .setChromeOptions(new chrome.Options().headless())
+      /*
+       * When running test in parallel:
+       * Without this line, even though all assertions pass, afterEach.always hook report intermittent errors.
+       * Usually first one or two tests execute successfully where rest of the tests raise errors.
+       * These erros are "ECONNREFUSED connect ECONNREFUSED 127.0.0.1:XXXXX" which are coming from webdriver.
+       * beforeEach hook creates a new driver for every test and they are
+       * supposed to run in isolation without affecting each other. At this point my best guess is that
+       * when driver.quit() is called, webdriver shuts down ChromeDriver server process thus causing
+       * subsequent requests to ChromeDriver server process fails.
+       *
+       * Further investigation of the issue lead me to following consequences;
+       * If we don't explicitly define ServiceBuilder for ChromeDriver it uses default ServiceBuilder
+       * which is a singleton, shared amongst different driver instances. As a result same
+       * ChromeDriver server process is used by different drivers. When driver.quit() is called by
+       * one of divers that process is terminated. By explicitly defining a new ServiceBuilder
+       * here we enforce creation of seperate ChromeDriver server child processes for each driver instance.
+       *
+       * If it is not feasible to create seperate ChromeDriver server child processes for each driver;
+       * instead of driver.quit(), driver.close() can be used in afterEach.always hook and
+       * driver.quit() can be called on after.always hook.
+       */
+      .setChromeService(new chrome.ServiceBuilder())
```

Comment:
Again, why not remove this entirely, and make this change instead:
```
diff --git a/client/tests/util/test.js b/client/tests/util/test.js
index 0361b93b9..37d819562 100644
--- a/client/tests/util/test.js
+++ b/client/tests/util/test.js
@@ -380,8 +380,16 @@ test.beforeEach(t => {
   }
 })
 
-// Shut down the browser when we are done.
+// Close the driver
 test.afterEach.always(async t => {
+  if (t.context.driver) {
+    await t.context.driver.close()
+  }
+})
+
+// The driver is shared among the parallel sessions,
+// so only shut down the browser when we are done with all tests.
+test.after.always(async t => {
   if (t.context.driver) {
     await t.context.driver.quit()
   }```
?

Reply:
Honestly, I haven't tested the change you propose. I have done all of my tests with current change and it works just fine. Also I choose first approach, because it is less code. I can't see any practical difference between two of them. But if you insist on the second approach, it is fine to me. I can test and make the change.

Reply:
In that case, at least rewrite/condense the extensive comment to just the factual reason for using a separate service on each driver.